### ISSUE TITLE
bug(fmols): align endogeneity and serial-correlation corrections with standard FMOLS

### ISSUE THREAD

**Issue by shawcharles** [sheep-farm/Greeners#23]
## Summary

Hayashi delegates `fmols()` directly to `greeners::fmols::FMOLS::fit`. A local, fixed-contract comparison shows that the current Greeners implementation agrees superficially with standard FMOLS on an IID control, but diverges when the cointegration error is serially correlated.

This is a Greeners estimator-correctness issue, not a request to widen Hayashi validation tolerances.

## Implementation evidence

At Greeners `dev` revision `b7a34ec4d8f1fcf1bfb7e5dc9bd83b6ec2af438c`:

- `crates/greeners-ols/src/fmols.rs:185-215` estimates only a symmetric long-run covariance and applies `Omega_ux Omega_xx^-1 Delta x` to `y`. It does not compute or apply the one-sided long-run covariance/bias correction used by standard Phillips-Hansen implementations.
- `crates/greeners-ols/src/fmols.rs:225-230` forms parameter covariance from unconditional `Omega_uu`. Standard implementations use the conditional long-run variance after projecting on the integrated regressors.
- Hayashi calls this implementation unchanged at `src/lang/interpreter/estimators_micro/panel.rs:390`.

## Reproducible evidence

All results use `n = 300`, intercept plus one I(1) regressor, a Bartlett kernel, and fixed bandwidth `5`. R uses `cointReg::cointRegFM(..., deter = intercept, kernel = "ba", bandwidth = 5)`; Python uses `arch.unitroot.cointegration.FullyModifiedOLS(..., trend = "c").fit(kernel = "bartlett", bandwidth = 5, force_int = TRUE)`.

### IID control

The existing `validation/cases/fmols_simulated` DGP gives close-looking results, which is why it cannot validate the correction:

| implementation | slope | slope SE |
| --- | ---: | ---: |
| R cointReg | 0.521140 | 0.064165 |
| Python arch | 0.520656 | 0.063267 |
| Hayashi/Greeners | 0.516540 | 0.063240 |

### Endogeneity stress

With seed `314159`, `dx ~ N(0,1)`, `x = cumsum(dx)`, `u = 0.75 * dx + e`, and `y = 0.25 + 0.5 * x + u`:

| implementation | slope | slope SE |
| --- | ---: | ---: |
| R cointReg | 0.526871 | 0.014211 |
| Python arch | 0.527363 | 0.014105 |
| Hayashi/Greeners | 0.524294 | 0.016012 |

### Serial-correlation stress

With seed `271828`, `dx ~ N(0,1)`, `x = cumsum(dx)`, `u_t = 0.7 u_(t-1) + e_t`, stationary initial `u_0`, and `y = 0.25 + 0.5 * x + u`:

| implementation | slope | slope SE |
| --- | ---: | ---: |
| R cointReg | 0.576868 | 0.045974 |
| Python arch | 0.578170 | 0.048303 |
| Hayashi/Greeners | 0.602853 | 0.050051 |

R and Python differ modestly but consistently form a convention-matched reference pair. Greeners differs by about `0.026` in the slope on the serial-correlation stress case, far beyond normal implementation variation for the stated contract.

## Proposed resolution

1. Confirm the intended contract is standard Phillips-Hansen FMOLS with an intercept, Bartlett kernel, and the existing automatic bandwidth default.
2. Implement the one-sided long-run covariance and the corresponding coefficient bias correction.
3. Use the conditional long-run variance for coefficient covariance and standard errors.
4. Add focused Greeners regression tests covering the IID, endogeneity, and serial-correlation DGPs above.
5. Only after the estimator is corrected, update Hayashi's `fmols_simulated` case with R and Python references and tolerances derived from reference-to-reference agreement.

Until then, Hayashi FMOLS should remain `not-supported`; matching only the IID control is behavioural-proxy evidence, not numerical validation.

**Comment by shawcharles** [comment]
Correction to my original comparison: **equal numeric bandwidth arguments did not give equal Bartlett weights**. R `cointReg` 0.2.0 at bandwidth 5 uses lag-1 to lag-5 weights `(0.8, 0.6, 0.4, 0.2, 0)`. Python `arch` 8.0.0 at bandwidth 5 uses `(5/6, 4/6, 3/6, 2/6, 1/6)`. **R bandwidth 6 matches arch bandwidth 5**. Calling the original pair convention-matched was incorrect. I also withdraw "far beyond normal implementation variation": these examples do not establish a general variation bound.

### Corrected Historical Results

These are native estimates recorded on 8 September 2026, not results from a rerun against today's Greeners or Hayashi. Each cell is **slope (slope standard error)**, with an intercept and 300 complete observations.

| Fixture | R cointReg 0.2.0, bandwidth 6 | arch 8.0.0, bandwidth 5 | Historical Hayashi executable |
| --- | ---: | ---: | ---: |
| IID | 0.520654846421 (0.063266903009) | 0.520655837078 (0.063266903009) | 0.516539661611 (0.063239805825) |
| Endogenous error | 0.527373213081 (0.014104820985) | 0.527363096579 (0.014104820985) | 0.524293782776 (0.016012354677) |
| Serially correlated error | 0.578090683556 (0.048302827540) | 0.578170338194 (0.048302827540) | 0.602853025772 (0.050050917894) |

The reference environment was Python 3.12.11, NumPy 2.3.5, pandas 2.3.3, arch 8.0.0, and R 4.6.1 with cointReg 0.2.0. The Hayashi executable reported `hay 0.2.11-dev`, SHA-256 `39120d82e7dbeff2899480648e88c31f52d60fd5fd2667afdbce140d484f40e4`. Its associated checkout was [Hayashi 63405fc](https://github.com/sheep-farm/hayashi/commit/63405fc5430776448dd42a706c4944ab4bfb25a8), whose lockfile named [Greeners b7a34ec](https://github.com/sheep-farm/Greeners/commit/b7a34ec4d8f1fcf1bfb7e5dc9bd83b6ec2af438c). That identifies the recorded executable and nearby source state; it does **not** prove that this binary was built from those sources. No clean build or current-version binary provenance is claimed.

### Finite-Sample Distinctions

- Let `N=300` and `n=N-1=299`. Both references use N observations for initial OLS and n for the FM regression and joint heteroskedasticity and autocorrelation consistent (HAC) covariance. The cited Greeners source retains N FM regression observations but uses n HAC observations. arch's N final residuals are not its FM regression sample size; R stores N residual slots with a leading NA.
- The matched references use a constant, no covariance demeaning, and no covariance degrees-of-freedom adjustment. Their one-sided covariance matrices have opposite orientation: `Delta_R = Lambda_arch.T`, for variable order `(u, dx)`. Both include the contemporaneous covariance term.
- R uses bias multiplier N; arch uses n. For slope-first design `Z=(x,1)` on rows 2 through N, put `c=Omega_ux/Omega_xx`, `b=Lambda_ux-c*Lambda_xx`, and `q=(b,0)'`. Their relationship is `beta_R = beta_arch - solve(Z'Z, q)`. This algebraic transformation predicted both native R coefficients to within `4.45e-16` in the recorded results. It is a diagnostic, **not a third independent estimator**, and is not substituted into the native table.
- Both reference covariance matrices use `(Omega_uu-Omega_ux^2/Omega_xx) * inverse(Z'Z)`. Their recorded standard errors agreed within `8.33e-17`. Those residuals describe this reconciliation, not acceptable error bounds for Greeners.

At the [cited Greeners source revision](https://github.com/sheep-farm/Greeners/blob/b7a34ec4d8f1fcf1bfb7e5dc9bd83b6ec2af438c/crates/greeners-ols/src/fmols.rs#L185-L230), the implementation adjusts the dependent variable using the symmetric long-run covariance, but does not compute and subtract the one-sided coefficient bias term. It uses unconditional `Omega_uu` for coefficient covariance, retains the first FM regression observation, and regularises inverses with small ridge terms. These are distinct differences: the observed coefficient gap does not identify the contribution of the omitted bias correction alone. That attribution would require controlled comparisons holding the other choices fixed.

The maintainer should choose and document a coherent finite-sample contract, including sample handling, bias multiplier, bandwidth weights, and covariance scaling. This comparison does not select R's convention over arch's. These three fixed datasets are diagnostic examples, not a Monte Carlo study or evidence about repeated-sampling bias, coverage, or size. They do not justify promoting Hayashi FMOLS from `not-supported` or setting its acceptance tolerance.

<details>
<summary>Standalone reproduction: shared fixtures, native references, and a separate algebraic diagnostic</summary>

Run the commands in a scratch directory. They create three CSV files consumed by both reference packages. Use Python with `numpy==2.3.5`, `pandas==2.3.3`, and `arch==8.0.0`, and R 4.6.1 with `cointReg` 0.2.0 installed. If using a separate R library, set `R_LIBS_USER` to its location before running Rscript. These are package calls and fixture generation, not copied estimator implementations. The diagnostic uses arch's version-sensitive private `_common_fit` API solely to obtain the joint covariance; `fit.kernel_est` concerns final residuals and is not a substitute.

```bash
python -B - <<'PY'
import hashlib
import platform
import numpy as np
import pandas as pd
import arch
from arch.unitroot.cointegration import FullyModifiedOLS

assert (np.__version__, pd.__version__, arch.__version__) == ("2.3.5", "2.3.3", "8.0.0")
print("Python", platform.python_version(), "numpy", np.__version__,
      "pandas", pd.__version__, "arch", arch.__version__)
N = 300
rng = np.random.RandomState(42)
x = rng.normal(0, 0.1, N).cumsum()
fixtures = {"iid": pd.DataFrame({"x": x, "y": 0.5*x + rng.normal(0, 0.5, N)})}
for name, seed in [("endogeneity", 314159), ("serial_correlation", 271828)]:
    rng = np.random.default_rng(seed)
    dx, e = rng.normal(size=N), rng.normal(size=N)
    x = dx.cumsum()
    u = 0.75*dx + e
    if name == "serial_correlation":
        u = np.empty(N)
        u[0] = e[0]/np.sqrt(1-0.7**2)
        for t in range(1, N):
            u[t] = 0.7*u[t-1] + e[t]
    fixtures[name] = pd.DataFrame({"x": x, "y": 0.25 + 0.5*x + u})

for name, frame in fixtures.items():
    path = name + ".csv"
    frame.to_csv(path, index=False, float_format="%.17g")
    with open(path, "rb") as stream:
        print(name, "CSV SHA-256", hashlib.sha256(stream.read()).hexdigest())
    df = pd.read_csv(path, float_precision="round_trip")
    model = FullyModifiedOLS(df.y, df[["x"]], trend="c")
    fit = model.fit(kernel="bartlett", bandwidth=5, force_int=True,
                    diff=False, df_adjust=False)
    print(name, "arch native: intercept, slope, SE(intercept), SE(slope)")
    print([float(v) for v in [fit.params["const"], fit.params["x"],
                              fit.std_errors["const"], fit.std_errors["x"]]])
    kernel, eta, _ = model._common_fit("bartlett", 5, True, False)
    assert eta.shape == (N-1, 2)
    omega = np.asarray(kernel.cov.long_run)
    lam = np.asarray(kernel.cov.one_sided)
    b = lam[0, 1] - omega[0, 1]/omega[1, 1]*lam[1, 1]
    z = np.column_stack([df.x.to_numpy()[1:], np.ones(N-1)])
    predicted_r = fit.params[["x", "const"]].to_numpy() - np.linalg.solve(z.T@z, [b, 0])
    print(name, "diagnostic predicted R: intercept, slope", predicted_r[::-1].tolist())
    print(name, "arch weights lag 0..5", kernel.kernel_weights.tolist())
PY

Rscript --vanilla - <<'RS'
stopifnot(as.character(getRversion()) == "4.6.1")
stopifnot(as.character(packageVersion("cointReg")) == "0.2.0")
suppressPackageStartupMessages(library(cointReg))
cat(R.version.string, "; cointReg", as.character(packageVersion("cointReg")), "\n")
options(digits = 16)
for (name in c("iid", "endogeneity", "serial_correlation")) {
  df <- read.csv(paste0(name, ".csv"))
  stopifnot(identical(names(df), c("x", "y")), nrow(df) == 300L,
            all(is.finite(as.matrix(df))))
  for (bw in c(5, 6)) {
    fit <- cointRegFM(as.matrix(df["x"]), df$y,
                     deter = matrix(1, nrow(df), 1), kernel = "ba",
                     bandwidth = bw, demeaning = FALSE)
    cat(name, "R native bandwidth", bw,
        ": intercept, slope, SE(intercept), SE(slope)\n")
    print(unname(c(fit$theta, fit$sd.theta)))
  }
}
sessionInfo()
RS
```

R bandwidth 5 above recreates the original mismatched reference; bandwidth 6 is the corrected comparison. Compare the independently obtained native R bandwidth-6 output with Python's explicitly labelled prediction. The prediction itself is not a native R result.

Optional Hayashi comparison, using an independently identified FMOLS-capable executable: the following `fmols.hay` reads the serial-correlation CSV and exports the slope and intercept with their standard errors. Substitute either other fixture filename to compare it. No bandwidth option is supplied; inspect the exported automatic bandwidth (historically 5).

```text
load "serial_correlation.csv" as df
let m = fmols(y ~ x, df)
let coefficients = select(m.coefficients, "variable", "coef", "std_err")
export(coefficients, "csv", "hay-coefficients.csv")
let fit = dataframe({
  "alpha": [m.alpha],
  "alpha_se": [m.alpha_se],
  "n_obs": [m.n_obs],
  "bandwidth": [m.bandwidth]
})
export(fit, "csv", "hay-fit.csv")
```

```bash
hay --version
sha256sum "$(command -v hay)"
hay fmols.hay
```

A version string and executable hash identify that run, not its source-to-binary provenance. A new executable may produce different results; these commands do not reproduce the historical executable or establish agreement with it.

</details>
